gufe/base.py

```diff
@@ -0,0 +1,381 @@
+    if (modname, qualname) in remappings:
+        modname, qualname = remappings[(modname, qualname)]
+
+    result = importlib.import_module(modname)
+    for name in qualname.split('.'):
+        result = getattr(result, name)
+
+    return result
+
+
+def get_class(module: str, qualname: str):
+    key = module, qualname
+    try:
+        return TOKENIZABLE_CLASS_REGISTRY[key]
+    except KeyError:
+        cls = import_qualname(module, qualname, REMAPPED_CLASSES)
+        TOKENIZABLE_CLASS_REGISTRY[key] = cls
+        return cls
+
+
+def modify_dependencies(obj: Union[Dict, List], modifier, is_mine, top=True):
+    """
+    Parameters
+    ----------
+    obj : Dict or List
+        Dictionary to traverse. Assumes that only mappings are dict and only
+        iterables are list, and that no gufe objects are in the keys of
+        dicts
+    modifier : Callable[[GufeTokenizable], Any]
+        function that modifies any GufeTokenizable found
+    is_mine : Callable[Any, bool]
+        function that determines whether the given object should be
+        subjected to the modifier
+    top : bool
+        If `True`, skip modifying `obj` itself; needed for recursive use to
+        avoid early stopping on `obj`.
+    """
+    obj = copy.deepcopy(obj)
+
+    if is_mine(obj) and not top:
+        obj = modifier(obj)
+
+    if isinstance(obj, dict):
+        obj = {key: modify_dependencies(value, modifier, is_mine, top=False)
+               for key, value in obj.items()}
+
+    elif isinstance(obj, list):
+        obj = [modify_dependencies(item, modifier, is_mine, top=False)
+               for item in obj]
+
+    return obj
+
+
+# encode options
+def to_dict(obj: GufeTokenizable) -> Dict:
+    dct = obj._to_dict()
+    dct.update(module_qualname(obj))
+    return dct
+
+
+def dict_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        to_dict,
+        is_gufe_obj,
+        top=True
+    )
+
+
+def key_encode_dependencies(obj: GufeTokenizable) -> Dict:
+    return modify_dependencies(
+        obj.to_shallow_dict(),
+        lambda obj: obj.key.to_dict(),
+        is_gufe_obj,
+        top=True
+    )
+
+
+# decode options
+def from_dict(dct) -> GufeTokenizable:
+    dct = copy.deepcopy(dct)
+
+    for key, val in dct.items():
+        if is_gufe_dict(val):
+            dct[key] = from_dict(val)
+
+    obj = _from_dict(dct)
+    try:
+        thing = TOKENIZABLE_REGISTRY[obj.key]()
+        # weakref will return None if the object was deleted
+        if thing is None:
+            return obj
+        else:
+            return thing
+    except KeyError:
+        return obj
+
+
+def _from_dict(dct: Dict) -> GufeTokenizable:
+    module = dct.pop('__module__')
+    qualname = dct.pop('__qualname__')
+    if (qualname is None) or (module is None):
+        raise KeyError("`__qualname__` or `__module__` key not found; unable "
```

Comment:
In principle, this should be impossible to reach. This code should only be called if something was tested positive for `is_gufe_dict`, which requires the presence of these keys.

Again, good defensive programming: I vote keep the code but mark it no-cover.

Reply:
Second thought here: I think it makes sense to have this logic, but to move it lower in the stack. I think this logic should be in `import_qualname` and tested as part of that.

Reply:
Agreed, thanks for doing this!